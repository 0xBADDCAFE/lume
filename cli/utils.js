import { existsSync } from "../deps/fs.js";
import lume from "../mod.js";
import { join, resolve } from "../deps/path.js";
import { brightGreen } from "../deps/colors.js";

/**
 * @return {Promise<*>} a lume instance - ready to build, run, etc.
 */
export async function buildSite(options, site) {
  options.root = resolve(Deno.cwd(), options.root);
  const configFile = join(options.root, options.config);

  if (!site) {
    if (existsSync(configFile)) {
      const mod = await import(`file://${configFile}`);
      site = mod.default;
    } else {
      site = lume();
    }
  }

  site.options.cwd = options.root;

  if (options.dev) {
    site.options.dev = options.dev;
  }

  if (options.location) {
    site.options.location = new URL(options.location);
  }

  if (options.src) {
    site.options.src = options.src;
  }

  if (options.dest) {
    site.options.dest = options.dest;
  }

  if (options["--"]) {
    site.options.flags = options["--"];
  }

  return site;
}

/**
 * @param command name of the command you're validating
 * @param args array of cli arguments (i.e. options._)
 * @param max the max number of args
 * @param min the min number of args
 */
export function validateArgsCount(command, args, max, min = 0) {
  if (args.length > max) {
    throw new Error(`Unexpected arguments: ${args.slice(max).join(", ")}
    
    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }

  if (args.length < min) {
    throw new Error(`Missing arguments

    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }
}
