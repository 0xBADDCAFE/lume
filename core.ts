import type {
  ComponentsOptions,
  default as Site,
  ServerOptions,
  SiteEvent,
  SiteEventType,
  SiteOptions,
  WatcherOptions,
} from "./core/site.ts";

import type {
  default as Events,
  Event,
  EventListener,
  EventOptions,
} from "./core/events.ts";

import type {
  default as Scripts,
  ScriptOptions,
  ScriptOrFunction,
} from "./core/scripts.ts";

import type {
  default as Reader,
  DirEntry,
  FileInfo,
  Loader,
} from "./core/reader.ts";
import type Logger from "./core/logger.ts";
import type Writer from "./core/writer.ts";
import type IncludesLoader from "./core/includes_loader.ts";
import type { default as PageLoader } from "./core/page_loader.ts";
import type DataLoader from "./core/data_loader.ts";
import type ComponentLoader from "./core/component_loader.ts";
import type {
  Component,
  Components,
  Content,
  Data,
  Dest,
  Directory,
  Page,
  Src,
  StaticFile,
} from "./core/filesystem.ts";
import type {
  ComponentFunction,
  default as Source,
  ProxyComponents,
} from "./core/source.ts";
import type {
  default as Renderer,
  Engine,
  Helper,
  HelperOptions,
} from "./core/renderer.ts";
import type { default as Processors, Processor } from "./core/processors.ts";
import type { default as Scopes, ScopeFilter } from "./core/scopes.ts";
import type { ErrorData, Exception } from "./core/errors.ts";
import type { default as Formats, Format } from "./core/formats.ts";
import type {
  default as Server,
  Middleware,
  RequestHandler,
  ServerEvent,
  ServerEventType,
} from "./core/server.ts";
import type {
  default as Watcher,
  WatchEvent,
  WatchEventType,
} from "./core/watcher.ts";

// Plugins types
import type { PaginationInfo, Paginator } from "./plugins/paginate.ts";
import type { Transformation, Transformations } from "./plugins/imagick.ts";
import type { MetaData } from "./plugins/metas.ts";
import type { Search } from "./plugins/search.ts";

/** The method that installs a plugin */
type PluginSetup = (options: unknown) => Plugin;

/** A generic Lume plugin */
type Plugin = (site: Site) => void;

export type {
  Component,
  ComponentFunction,
  ComponentLoader,
  Components,
  ComponentsOptions,
  Content,
  Data,
  DataLoader,
  Dest,
  Directory,
  DirEntry,
  Engine,
  ErrorData,
  Event,
  EventListener,
  EventOptions,
  Events,
  Exception,
  FileInfo,
  Format,
  Formats,
  Helper,
  HelperOptions,
  IncludesLoader,
  Loader,
  Logger,
  Middleware,
  Page,
  PageLoader,
  Plugin,
  PluginSetup,
  Processor,
  Processors,
  ProxyComponents,
  Reader,
  Renderer,
  RequestHandler,
  ScopeFilter,
  Scopes,
  ScriptOptions,
  ScriptOrFunction,
  Scripts,
  Server,
  ServerEvent,
  ServerEventType,
  ServerOptions,
  Site,
  SiteEvent,
  SiteEventType,
  SiteOptions,
  Source,
  Src,
  StaticFile,
  Watcher,
  WatcherOptions,
  WatchEvent,
  WatchEventType,
  Writer,
};

export interface PageData extends Data {
  /** The title of the page */
  title?: string;

  /** The page url */
  url: string;

  /** The available components */
  // deno-lint-ignore no-explicit-any
  comp?: any;

  /** Plugin Paginator: The paginator helper */
  paginate?: Paginator<Page>;

  /** Plugin Pagination: The pagination info */
  pagination?: PaginationInfo;

  /** Plugin Pagination: The pagination result */
  results?: Page[];

  /** Plugin Imagick: Image transformations */
  imagick?: Transformation | Transformations;

  /** Plugin Metas: Meta elements */
  metas?: MetaData;

  /** Plugin NetlifyCMS: CMS configuration */
  netlify_cms?: Record<string, unknown>;

  /** Plugin Search: The searcher helper */
  search?: Search;
}
