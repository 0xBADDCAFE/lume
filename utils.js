import { bold, red } from "./deps/colors.js";
import { SEP } from "./deps/path.js";

export async function concurrent(iterable, iteratorFn, limit = 200) {
  const executing = [];

  for await (const item of iterable) {
    const p = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

export const mimes = new Map([
  [".html", "text/html; charset=utf-8"],
  [".txt", "text/plain; charset=utf-8"],
  [".js", "text/javascript; charset=utf-8"],
  [".css", "text/css; charset=utf-8"],
  [".atom", "application/atom+xml; charset=utf-8"],
  [".rss", "application/rss+xml; charset=utf-8"],
  [".svg", "image/svg+xml; charset=utf-8"],
  [".json", "application/json"],
  [".webmanifest", "application/manifest+json"],
  [".ico", "image/x-icon"],
  [".avif", "image/avif"],
  [".png", "image/png"],
  [".apng", "image/apng"],
  [".jpg", "image/jpg"],
  [".jpeg", "image/jpg"],
  [".gif", "image/gif"],
  [".mp3", "audio/mpeg"],
  [".mp4", "video/mpeg"],
  [".xml", "text/xml"],
  [".woff", "font/woff"],
  [".woff2", "font/woff2"],
  [".wasm", "application/wasm"],
  [".webp", "image/webp"],
  [".webm", "video/webm"],
  [".zip", "application/zip"],
]);

export function error(context, message, exception) {
  console.error(bold(red(`${context}:`)), message);
  if (exception) {
    console.error(exception);
  }
  console.error("");
}

export function merge(defaults, user) {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    if (
      typeof merged[key] === "object" && typeof value === "object" &&
      !Array.isArray(merged[key])
    ) {
      merged[key] = merge(merged[key], value);
      continue;
    }

    merged[key] = value;
  }

  return merged;
}

export function normalizePath(path) {
  // Is Windows path
  if (SEP !== "/") {
    return path.replaceAll(SEP, "/");
  }

  return path;
}

export const slugChars = new Map([
  ["รฐ", "d"],
  ["ร", "ss"],
]);

export function slugify(string) {
  return string.toLowerCase()
    .normalize("NFKD")
    .replaceAll(/[\s_-]+/g, "-")
    .replaceAll(/[^\w\/-]/g, (char) => slugChars.get(char) || "");
}

export function searchByExtension(path, extensions) {
  for (const [key, value] of extensions) {
    if (path.endsWith(key)) {
      return [key, value];
    }
  }
}
