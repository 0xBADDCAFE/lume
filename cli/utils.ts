import lume from "../mod.ts";
import { toFileUrl } from "../deps/path.ts";
import { dim, red } from "../deps/colors.ts";
import { getConfigFile } from "../core/utils.ts";

import type { Site } from "../core.ts";

/** Create a site instance */
export async function createSite(root: string, config?: string): Promise<Site> {
  const path = await getConfigFile(root, config);

  if (path) {
    console.log(`Loading config file ${dim(path)}`);
    console.log();
    const mod = await import(toFileUrl(path).href);
    return mod.default;
  }

  return lume();
}

/** A list of the available optional plugins */
export const pluginNames = [
  "attributes",
  "base_path",
  "code_highlight",
  "date",
  "esbuild",
  "eta",
  "imagick",
  "inline",
  "jsx",
  "lightningcss",
  "liquid",
  "metas",
  "minify_html",
  "modify_urls",
  "multilanguage",
  "nano_jsx",
  "netlify_cms",
  "on_demand",
  "postcss",
  "prism",
  "pug",
  "relations",
  "relative_urls",
  "resolve_urls",
  "sass",
  "slugify_urls",
  "svgo",
  "terser",
  "windi_css",
];

export function log(...lines: (string | undefined)[]) {
  console.log("----------------------------------------");
  lines.forEach((line) => line && console.log(line));
  console.log("----------------------------------------");
}

export function promptConfigUpdate(data: unknown) {
  log(
    red("deno.jsonc needs to be manually updated:"),
    dim("Use deno.json to update it automatically"),
    JSON.stringify(data, null, 2),
  );
}
