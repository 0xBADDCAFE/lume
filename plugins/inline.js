import { extname, join, relative, resolve } from "../deps/path.js";
import { DOMParser } from "../deps/dom.js";
import { encode } from "../deps/base64.js";
import { mimes } from "../utils.js";

const cache = new Map();

export default function () {
  const parser = new DOMParser();

  return (site) => {
    site.process([".html"], processor);

    //Update cache
    site.addEventListener("beforeUpdate", (ev) => {
      for (const filename of ev.files) {
        cache.delete(filename);
      }
    });

    async function processor(page) {
      if (page.content.includes(" inline")) {
        const document = parser.parseFromString(page.content, "text/html");

        for (const element of document.querySelectorAll("[inline]")) {
          await inline(page.data.url, element);
          // bug: https://github.com/b-fuze/deno-dom/issues/30
          // element.removeAttribute("inline");
          delete element.attributes["inline"];
        }

        page.content = document.documentElement.outerHTML;
      }
    }

    async function inline(url, element) {
      if (element.hasAttribute("href")) {
        if (element.getAttribute("rel") === "stylesheet") {
          return inlineStyles(url, element);
        }

        return inlineHref(url, element);
      }

      if (element.hasAttribute("src")) {
        if (element.nodeName === "SCRIPT") {
          return inlineScript(url, element);
        }

        return inlineSrc(url, element);
      }
    }

    async function getContent(path, asData = false) {
      //Ensure the path starts with "/"
      path = join("/", path);

      if (cache.has(path)) {
        return cache.get(path);
      }

      const content = await readContent(path, asData);
      cache.set(path, content);
      return content;
    }

    async function readContent(path, asData) {
      const url = join("/", relative(site.options.location.pathname, path));

      //Is a page/asset ?
      const page = site.pages.find((page) => page.data.url === url);

      if (page) {
        return page.content;
      }

      //Is a file in dest
      if (!asData) {
        return Deno.readTextFile(site.dest(url));
      }

      const content = await Deno.readFile(site.dest(url));
      const ext = extname(path);

      if (!mimes.has(ext)) {
        throw new Error(`Unknown mime type for ${path}`);
      }

      return `data:${mimes.get(ext)};base64,${encode(content)}`;
    }

    async function inlineStyles(url, element) {
      const path = resolve(url, element.getAttribute("href"));
      const style = element.ownerDocument.createElement("style");

      style.innerHTML = await getContent(path);
      element.replaceWith(style);
    }

    async function inlineScript(url, element) {
      const path = resolve(url, element.getAttribute("src"));

      element.innerHTML = await getContent(path);
      // bug: https://github.com/b-fuze/deno-dom/issues/30
      // element.removeAttribute("src");
      delete element.attributes["src"];
    }

    async function inlineSrc(url, element) {
      const path = resolve(url, element.getAttribute("src"));
      const ext = extname(path);

      if (ext === ".svg") {
        const content = await getContent(path);
        const div = element.ownerDocument.createElement("div");
        div.innerHTML = content;
        element.replaceWith(...div.children);
        return;
      }

      element.setAttribute("src", await getContent(path, true));
    }

    async function inlineHref(url, element) {
      const path = resolve(url, element.getAttribute("href"));
      element.setAttribute("href", await getContent(path, true));
    }
  };
}
