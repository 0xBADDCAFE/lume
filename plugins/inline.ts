import { Element } from "../deps/dom.ts";
import { extname, posix, resolve } from "../deps/path.ts";
import { encode } from "../deps/base64.ts";
import { merge, mimes, warn } from "../core/utils.ts";
import { Page, Site } from "../core.ts";
import binaryLoader from "../core/loaders/binary.ts";

export interface Options {
  /** The list of extensions this plugin applies to */
  extensions: string[];

  /** Attribute used to select the elements this plugin applies to */
  attribute: string;
}

// Default options
const defaults: Options = {
  extensions: [".html"],
  attribute: "inline",
};

const cache = new Map();

/**
 * A plugin to inline the HTML assets,
 * like images, JavaScript, CSS, SVG, etc.
 */
export default function (userOptions?: Partial<Options>) {
  const options = merge(defaults, userOptions);

  return (site: Site) => {
    site.process(options.extensions, inline);

    // Update the cache
    site.addEventListener("beforeUpdate", (ev) => {
      for (const filename of ev.files!) {
        cache.delete(filename);
      }
    });

    const selector = `[${options.attribute}]`;

    async function inline(page: Page) {
      for (const element of page.document!.querySelectorAll(selector)) {
        await runInline(page.data.url as string, element as Element);
        (element as Element).removeAttribute(options.attribute);
      }
    }

    function runInline(url: string, element: Element) {
      if (element.hasAttribute("href")) {
        return element.getAttribute("rel") === "stylesheet"
          ? inlineStyles(url, element)
          : inlineHref(url, element);
      }

      if (element.hasAttribute("src")) {
        return element.nodeName === "SCRIPT"
          ? inlineScript(url, element)
          : inlineSrc(url, element);
      }
    }

    function getContent(path: string, asDataUrl = false) {
      // Ensure the path starts with "/"
      path = posix.join("/", path);
      const id = JSON.stringify([path, asDataUrl]);

      if (!cache.has(id)) {
        cache.set(id, readContent(path, asDataUrl));
      }

      return cache.get(id);
    }

    async function readContent(path: string, asDataUrl: boolean) {
      const url = posix.join(
        "/",
        posix.relative(site.options.location.pathname, path),
      );

      const content = await getFileContent(site, url) as string | Uint8Array;

      // Return the raw content
      if (!asDataUrl) {
        if (content instanceof Uint8Array) {
          return new TextDecoder().decode(content);
        }

        return content;
      }

      // Return the data URL
      const ext = extname(path);

      if (!mimes.has(ext)) {
        warn("Unknown file format", {
          name: "Inline plugin",
          path,
          url,
          available: mimes,
        });
      }

      return `data:${mimes.get(ext)};base64,${encode(content)}`;
    }

    async function inlineStyles(url: string, element: Element) {
      const path = posix.resolve(url, element.getAttribute("href")!);
      const style = element.ownerDocument!.createElement("style");

      try {
        style.innerHTML = await getContent(path);
        element.replaceWith(style);
      } catch (cause) {
        warn("Unable to inline the file", {
          name: "Inline plugin",
          cause,
          path,
          url,
        });
      }
    }

    async function inlineScript(url: string, element: Element) {
      const path = posix.resolve(url, element.getAttribute("src")!);

      try {
        element.innerHTML = await getContent(path);
        element.removeAttribute("src");
      } catch (cause) {
        warn("Unable to inline the file", {
          name: "Inline plugin",
          cause,
          path,
          url,
        });
      }
    }

    async function inlineSrc(url: string, element: Element) {
      const path = resolve(url, element.getAttribute("src")!);
      const ext = extname(path);

      try {
        if (ext === ".svg") {
          const content = await getContent(path);
          const div = element.ownerDocument!.createElement("div");
          div.innerHTML = content;
          element.replaceWith(...div.children);
          return;
        }

        element.setAttribute("src", await getContent(path, true));
      } catch (cause) {
        warn("Unable to inline the file", {
          name: "Inline plugin",
          cause,
          path,
          url,
        });
      }
    }

    async function inlineHref(url: string, element: Element) {
      const path = resolve(url, element.getAttribute("href")!);

      try {
        element.setAttribute("href", await getContent(path, true));
      } catch (cause) {
        warn("Unable to inline the file", {
          name: "Inline plugin",
          cause,
          path,
          url,
        });
      }
    }
  };
}

/** Returns the content of a file or page */
async function getFileContent(
  site: Site,
  url: string,
): Promise<string | Uint8Array> {
  // Is a loaded file
  const page = site.pages.find((page) => page.data.url === url);

  if (page) {
    return page.content as string | Uint8Array;
  }

  // Is a static file
  for (const entry of site.source.staticFiles) {
    const [from, to] = entry;

    if (url.startsWith(to)) {
      const file = site.src(from, url.slice(to.length));
      const content = await site.reader.read(file, binaryLoader);
      return content.content as Uint8Array;
    }
  }

  // Is a source file
  const content = await site.reader.read(site.src(url), binaryLoader);
  return content.content as Uint8Array;
}
