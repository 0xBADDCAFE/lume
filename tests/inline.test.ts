import { assert, assertStrictEquals as equals } from "../deps/assert.ts";
import { build, getSite, testPage } from "./utils.ts";
import inline from "../plugins/inline.ts";
import binaryLoader from "../core/loaders/binary.ts";
import { Page } from "../core.ts";

Deno.test("code_hightlight plugin", async () => {
  const site = getSite({
    src: "inline",
  });

  site.use(inline());
  site.loadAssets([".svg", ".js"]);
  site.loadAssets([".png"], binaryLoader);
  site.copy("favicon.png", "favicon2.png");
  await build(site);

  testPage(site, "/styles.css", (page) => {
    equals(page.data.url, "/styles.css");
  });

  testPage(site, "/index", (page) => {
    assert(!inlined(page, "el-1", "href"));
    equals(page.document?.querySelector("#el-1 + style")?.tagName, "STYLE");
    assert(inlined(page, "el-3", "href"));
    assert(!inlined(page, "el-4", "href"));
    assert(inlined(page, "el-5", "href"));
    assert(!inlined(page, "el-6", "href"));
    equals(page.document?.querySelector("#el-7")?.tagName, "SVG");
    equals(page.document?.querySelector("#el-7")?.className, "has-svg");
    assert(!inlined(page, "el-8", "src"));
    assert(inlined(page, "el-9", "src"));
    assert(!inlined(page, "el-10", "src"));
    assert(!inlinedScript(page, "el-11"));
    assert(inlinedScript(page, "el-12"));
    assert(inlinedScript(page, "el-13"));
  });
});

function inlined(page: Page, id: string, attr: string) {
  const href = page.document?.getElementById(id)?.getAttribute(attr);

  return href && href.startsWith("data:");
}

function inlinedScript(page: Page, id: string) {
  return page.document?.getElementById(id)?.innerHTML !== "";
}
