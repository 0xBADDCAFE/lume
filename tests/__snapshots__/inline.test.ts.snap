export const snapshot = {};

snapshot[`inline plugin 1`] = `4`;

snapshot[`inline plugin 2`] = `
{
  formats: [
    {
      engine: true,
      key: ".tmpl.ts",
      pageLoader: [AsyncFunction: module],
      removeExtension: true,
    },
    {
      engine: true,
      key: ".tmpl.js",
      pageLoader: [AsyncFunction: module],
      removeExtension: true,
    },
    {
      engine: false,
      key: ".tmpl.json",
      pageLoader: [AsyncFunction: json],
      removeExtension: true,
    },
    {
      dataLoader: [AsyncFunction: json],
      engine: false,
      key: ".json",
    },
    {
      engine: true,
      key: ".md",
      pageLoader: [AsyncFunction: text],
      removeExtension: true,
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engine: true,
      key: ".js",
      pageLoader: [AsyncFunction: text],
      removeExtension: false,
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engine: true,
      key: ".ts",
    },
    {
      componentLoader: [AsyncFunction: text],
      engine: true,
      includesPath: "_includes",
      key: ".njk",
      pageLoader: [AsyncFunction: text],
      removeExtension: true,
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engine: false,
      key: ".yaml",
      pageLoader: [AsyncFunction: yaml],
      removeExtension: true,
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engine: false,
      key: ".yml",
      pageLoader: [AsyncFunction: yaml],
      removeExtension: true,
    },
    {
      engine: false,
      key: ".svg",
      pageLoader: [AsyncFunction: text],
      removeExtension: false,
    },
    {
      engine: false,
      key: ".png",
      pageLoader: [AsyncFunction: binary],
      removeExtension: false,
    },
  ],
}
`;

snapshot[`inline plugin 3`] = `
[
  {
    dest: "/favicon2.png",
    saved: true,
    src: "/favicon.png",
  },
]
`;

snapshot[`inline plugin 4`] = `
{
  content: '<?xml version="1.0" encoding="UTF-8"?>
<svg width="126px" height="126px" viewBox="0 0 126 126" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <title>favicon</title>
    <g id="favicon" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <path d="M52.8683497,7.99450751 C48.9069025,24.0969979 69.1669662,21.1536885 69.1669662,44.7198348 C75.794917,37.1051328 70.4463322,26.8453748 81.058456,17.1725913 C81.058456,39.3863344 102.155357,38.1365645 94.0121919,69.2249916 C97.7866664,66.9118477 100.501055,63.103682 102.155357,57.8004944 C105.491398,63.103682 104.245257,72.5297487 98.416934,86.0786945 C103.547554,83.9150474 106.558165,80.9232746 108.378132,76.7271327 C109.04106,105.964869 89.5062156,126.366871 61.9867103,125.995028 C34.467205,125.623185 13.5784967,101.642443 18.2649214,71.802957 C18.9624768,74.9072731 21.5182769,78.5533566 25.1832626,80.0729101 C18.6891247,50.7321037 30.922502,49.6337285 36.0056643,33.9730813 C35.1796585,42.297059 36.0056643,46.924952 40.0208236,51.7593505 C40.7309276,49.6474958 42.0145668,47.6388953 43.3724725,45.627895 L43.7620351,45.0531669 C43.8271552,44.9573376 43.892336,44.86148 43.9575234,44.7655828 L44.5429346,43.9011285 C47.3268507,39.7628977 49.8240593,35.4880393 47.7482658,30.1694988 C44.3690672,21.5114095 44.6913719,18.1446262 52.8683497,7.99450751 Z M43.4168797,15.6028981 C41.2434833,17.8205255 41.7153274,24.275166 44.213695,29.9427578 C46.7120625,35.6103495 43.4168797,39.8756006 39.8854597,45.126676 C40.5019929,38.1390602 39.8587244,36.4802473 36.8588412,31.1077941 C33.858958,25.7353408 37.9685715,18.6230687 43.4168797,15.6028981 Z M64.7721431,8.03208837 C69.8550649,12.4485247 71.6576876,12.8901943 72.209291,17.4374541 C72.7608943,21.9847139 70.2967086,24.09635 71.2822168,30.0346208 C60.9002329,21.508497 69.0471936,18.4710785 64.7721431,8.03208837 Z M53.3587168,1.6350624e-05 C57.0113944,4.33013045 59.1813933,3.19675313 61.5638477,8.73089813 C63.9463021,14.2650431 60.9615342,15.5202483 63,22.1753596 C51.0029851,11.8529541 58.4461322,11.0173257 53.3587168,1.6350624e-05 Z" id="Combined-Shape-Copy" fill="#C84F46"></path>
    </g>
</svg>',
  data: {
    content: '<?xml version="1.0" encoding="UTF-8"?>
<svg width="126px" height="126px" viewBox="0 0 126 126" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <title>favicon</title>
    <g id="favicon" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <path d="M52.8683497,7.99450751 C48.9069025,24.0969979 69.1669662,21.1536885 69.1669662,44.7198348 C75.794917,37.1051328 70.4463322,26.8453748 81.058456,17.1725913 C81.058456,39.3863344 102.155357,38.1365645 94.0121919,69.2249916 C97.7866664,66.9118477 100.501055,63.103682 102.155357,57.8004944 C105.491398,63.103682 104.245257,72.5297487 98.416934,86.0786945 C103.547554,83.9150474 106.558165,80.9232746 108.378132,76.7271327 C109.04106,105.964869 89.5062156,126.366871 61.9867103,125.995028 C34.467205,125.623185 13.5784967,101.642443 18.2649214,71.802957 C18.9624768,74.9072731 21.5182769,78.5533566 25.1832626,80.0729101 C18.6891247,50.7321037 30.922502,49.6337285 36.0056643,33.9730813 C35.1796585,42.297059 36.0056643,46.924952 40.0208236,51.7593505 C40.7309276,49.6474958 42.0145668,47.6388953 43.3724725,45.627895 L43.7620351,45.0531669 C43.8271552,44.9573376 43.892336,44.86148 43.9575234,44.7655828 L44.5429346,43.9011285 C47.3268507,39.7628977 49.8240593,35.4880393 47.7482658,30.1694988 C44.3690672,21.5114095 44.6913719,18.1446262 52.8683497,7.99450751 Z M43.4168797,15.6028981 C41.2434833,17.8205255 41.7153274,24.275166 44.213695,29.9427578 C46.7120625,35.6103495 43.4168797,39.8756006 39.8854597,45.126676 C40.5019929,38.1390602 39.8587244,36.4802473 36.8588412,31.1077941 C33.858958,25.7353408 37.9685715,18.6230687 43.4168797,15.6028981 Z M64.7721431,8.03208837 C69.8550649,12.4485247 71.6576876,12.8901943 72.209291,17.4374541 C72.7608943,21.9847139 70.2967086,24.09635 71.2822168,30.0346208 C60.9002329,21.508497 69.0471936,18.4710785 64.7721431,8.03208837 Z M53.3587168,1.6350624e-05 C57.0113944,4.33013045 59.1813933,3.19675313 61.5638477,8.73089813 C63.9463021,14.2650431 60.9615342,15.5202483 63,22.1753596 C51.0029851,11.8529541 58.4461322,11.0173257 53.3587168,1.6350624e-05 Z" id="Combined-Shape-Copy" fill="#C84F46"></path>
    </g>
</svg>',
    date: 1970-01-01T00:00:00.000Z,
    paginate: [GeneratorFunction: paginate],
    search: Search {},
    tags: [
    ],
    url: "/favicon.svg",
  },
  dest: {
    ext: ".svg",
    path: "/favicon",
  },
  src: {
    ext: ".svg",
    path: "/favicon",
  },
}
`;

snapshot[`inline plugin 5`] = `
{
  content: '<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="UTF-8">
  <title>Inline test</title>
  <link id="el-1" rel="stylesheet" href="styles.css">
  <style>body {
  background: white;
  color: #333;
}</style>
  <link id="el-3" rel="shortcut icon" href="data:image/svg+xml; charset=utf-8;base64,(...)">
  <link id="el-4" rel="shortcut icon" href="/favicon.svg">
  <link id="el-5" rel="shortcut icon" href="data:image/png;base64,(...)">
  <link id="el-6" rel="shortcut icon" href="/favicon2.png">
</head>
<body>
  <svg width="126px" height="126px" viewbox="0 0 126 126" version="1.1" xmlns="http://www.w3.org/2000/svg" xlink="http://www.w3.org/1999/xlink" class="has-svg" id="el-7">
    <title>favicon</title>
    <g id="favicon" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <path d="M52.8683497,7.99450751 C48.9069025,24.0969979 69.1669662,21.1536885 69.1669662,44.7198348 C75.794917,37.1051328 70.4463322,26.8453748 81.058456,17.1725913 C81.058456,39.3863344 102.155357,38.1365645 94.0121919,69.2249916 C97.7866664,66.9118477 100.501055,63.103682 102.155357,57.8004944 C105.491398,63.103682 104.245257,72.5297487 98.416934,86.0786945 C103.547554,83.9150474 106.558165,80.9232746 108.378132,76.7271327 C109.04106,105.964869 89.5062156,126.366871 61.9867103,125.995028 C34.467205,125.623185 13.5784967,101.642443 18.2649214,71.802957 C18.9624768,74.9072731 21.5182769,78.5533566 25.1832626,80.0729101 C18.6891247,50.7321037 30.922502,49.6337285 36.0056643,33.9730813 C35.1796585,42.297059 36.0056643,46.924952 40.0208236,51.7593505 C40.7309276,49.6474958 42.0145668,47.6388953 43.3724725,45.627895 L43.7620351,45.0531669 C43.8271552,44.9573376 43.892336,44.86148 43.9575234,44.7655828 L44.5429346,43.9011285 C47.3268507,39.7628977 49.8240593,35.4880393 47.7482658,30.1694988 C44.3690672,21.5114095 44.6913719,18.1446262 52.8683497,7.99450751 Z M43.4168797,15.6028981 C41.2434833,17.8205255 41.7153274,24.275166 44.213695,29.9427578 C46.7120625,35.6103495 43.4168797,39.8756006 39.8854597,45.126676 C40.5019929,38.1390602 39.8587244,36.4802473 36.8588412,31.1077941 C33.858958,25.7353408 37.9685715,18.6230687 43.4168797,15.6028981 Z M64.7721431,8.03208837 C69.8550649,12.4485247 71.6576876,12.8901943 72.209291,17.4374541 C72.7608943,21.9847139 70.2967086,24.09635 71.2822168,30.0346208 C60.9002329,21.508497 69.0471936,18.4710785 64.7721431,8.03208837 Z M53.3587168,1.6350624e-05 C57.0113944,4.33013045 59.1813933,3.19675313 61.5638477,8.73089813 C63.9463021,14.2650431 60.9615342,15.5202483 63,22.1753596 C51.0029851,11.8529541 58.4461322,11.0173257 53.3587168,1.6350624e-05 Z" id="Combined-Shape-Copy" fill="#C84F46"></path>
    </g>
</svg>
  <img id="el-8" class="has-svg" src="./favicon.svg">
  <img id="el-9" class="has-png" src="data:image/png;base64,(...)">
  <img id="el-10" class="has-png" src="./favicon2.png">
  <script id="el-11" src="./script.js"></script>
  <script id="el-12">console.log("Hello world");
const a = 1, _b = 1 < a, _c = a > 4;
</script>
  <script id="el-13" type="module">console.log("Hello world");
const a = 1, _b = 1 < a, _c = a > 4;
</script>


</body></html>',
  data: {
    content: '<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Inline test</title>
  <link id="el-1" rel="stylesheet" href="styles.css">
  <link id="el-2" rel="stylesheet" href="styles.css" inline>
  <link id="el-3" rel="shortcut icon" href="/favicon.svg" inline>
  <link id="el-4" rel="shortcut icon" href="/favicon.svg">
  <link id="el-5" rel="shortcut icon" href="/favicon2.png" inline>
  <link id="el-6" rel="shortcut icon" href="/favicon2.png">
</head>
<body>
  <img id="el-7" class="has-svg" src="./favicon.svg" inline>
  <img id="el-8" class="has-svg" src="./favicon.svg">
  <img id="el-9" class="has-png" src="./favicon2.png" inline>
  <img id="el-10" class="has-png" src="./favicon2.png">
  <script id="el-11" src="./script.js"></script>
  <script id="el-12" src="./script.js" inline></script>
  <script id="el-13" src="./script.js" type="module" inline></script>
</body>
</html>
',
    date: 1970-01-01T00:00:00.000Z,
    paginate: [GeneratorFunction: paginate],
    search: Search {},
    tags: [
    ],
    url: "/",
  },
  dest: {
    ext: ".html",
    path: "/index",
  },
  src: {
    ext: ".njk",
    path: "/index",
  },
}
`;

snapshot[`inline plugin 6`] = `
{
  content: 'console.log("Hello world");
const a = 1, _b = 1 < a, _c = a > 4;
',
  data: {
    content: 'console.log("Hello world");
const a = 1, _b = 1 < a, _c = a > 4;
',
    date: 1970-01-01T00:00:00.000Z,
    paginate: [GeneratorFunction: paginate],
    search: Search {},
    tags: [
    ],
    url: "/script.js",
  },
  dest: {
    ext: ".js",
    path: "/script",
  },
  src: {
    ext: ".js",
    path: "/script",
  },
}
`;

snapshot[`inline plugin 7`] = `
{
  content: "body {
  background: white;
  color: #333;
}",
  data: {
    content: "body {
  background: white;
  color: #333;
}",
    date: 1970-01-01T00:00:00.000Z,
    paginate: [GeneratorFunction: paginate],
    search: Search {},
    tags: [
    ],
    url: "/styles.css",
  },
  dest: {
    ext: ".css",
    path: "/styles",
  },
  src: {
    ext: ".njk",
    path: "/styles.css",
  },
}
`;
